Add API endpoint to logout current user.

njohner opened the pull request:
We add an API endpoint to log out the current user by expiring his cookies (__ac and deployment specific cookie, e.g. __ac_fd). This is necessary because logout in the frontend should also log the user out of the backend.

I'm actually really not sure this is the right approach. It works, but maybe that should be the job of the portal or something?
I haven't added the documentation yet, as I'm more and more doubtful that this is a good solution.

For https://4teamwork.atlassian.net/browse/GEVER-558
PR in the frontend: https://github.com/4teamwork/gever-ui/pull/1186

## Checklist (Must have)

_Everything has to be done/checked. Checked but not present means the author deemed it unnecessary._

- [x] Changelog entry
- [x] Link to issue (Jira or GitHub) and backlink in issue (Jira)


## Checklist (if applicable)

- API change:
  - [ ] Documentation is updated

tinagerber commented:
https://plonerestapi.readthedocs.io/en/latest/conventions.html#naming-convention-for-attribute-names-in-uris

njohner commented:
right thanks...

elioschmutz commented on opengever/api/logout_user.py:
```diff
@@ -0,0 +1,19 @@
+
+        # Also expire any __ac cookie that might have been issued by the
+        # CookieAuthHelper on the Zope root (e.g. for zopemaster)
+        self.request.response.expireCookie('__ac', path='/')
```
can't we somehow reuse the `@@logout` view? instead of the redirect, they have to do the same thing

njohner replied:
I'd say no. It's basically a single line of code that is shared, and if we could work around the redirect it would be 2 lines of code. So I'd rather simply duplicate that here.

elioschmutz commented:
@njohner 👍 but I'm not sure either, if this is the right approach. So perhaps someone else should take another look to it...

deiferni commented on opengever/api/logout_user.py:
```diff
@@ -0,0 +1,19 @@
+        # Also expire any __ac cookie that might have been issued by the
+        # CookieAuthHelper on the Zope root (e.g. for zopemaster)
+        self.request.response.expireCookie('__ac', path='/')
+        return
```
should we do a 204 no-content status? or does that happen already by returning `None`?

njohner commented:
Ok I've now overwritten the `@logout` endpoint instead. I've also improved the tests to actually test that logout works and added a few new ones. I've also added some documentation for the endpoint.
As there was no real opposition, I'd propose we that we go with this for now. Once we only have the new UI, we will probably not need to invalidate the session cookies anymore and we can then remove that customisation.

buchi commented on opengever/api/logout.py:
```diff
@@ -0,0 +1,34 @@
+    def reply(self):
+        # Handles logout by invalidating the JWT
+        # This sets the response status accordingly if it fails
+        resp = super(GeverLogout, self).reply()
```
I see one problem now:
The frontend does not know if there's a `__ac` cookie, thus it always has to call the `@logout` endpoint. However if there's no `__ac` cookie, it will get a 501 response because by default we do not store JWT tokens and thus it's not possible to invalidate those on the server side. Usually this is not needed as the Client can just delete the token client-side to logout.
This can't be done with the `__ac` cookies as those should be `httponly` by default and thus the server needs to delete 'em.

Not sure how we want to deal with that. Maybe add a URL parameter `invalidate=false`. And if that's given, skip the JWT invalidation? 

`__ac` cookies can't be invalidated individually, just deleted.

njohner replied:
Yes currently the frontend gets a 501 but ignores it. I mean it deletes it's own session token and then basically tries to logout from the backend just in case, and that might fail, but it doesn't really care.
With the current implementation I would also return a 501 even if we skip deleting the JWT token if there is no `__ac` or customized (`__ac_fd`) cookie to delete. I sort of think this is OK, it's the backend answering that it could not logout.
